### ISSUE TITLE
Using CXX toolchain in custom rules (protobuf)

### ISSUE THREAD

**Issue by benbrittain** [facebook/buck2#207]
I'm trying to put together a quick protobuf toolchain for C++.

Generating the `*.{pb.cc, pb.h}` files is easy enough, but I'm not confident on the "best practice" for how to feed them into the C++ toolchain. I have a rule that generates a directory artifact that currently is returned through `DefaultInfo`.

Should I be using a macro that somehow feeds the result of that rule into the `srcs/headers` field of a `native.cxx_library`? (I'm not confident on the relationship between DefaultInfo and fields into rules)

Or should I be writing a `cxx_proto_library` rule, and if so is there a way of not re-implementing most of the logic contained withing `cxx_library`?

**Comment by dtolnay** [comment]
I think you want a macro for this, not (or in addition to) a rule. That's how we do Thrift.

This commit message has a glimpse of how Thrift works that might be educative: https://github.com/facebook/buck2/commit/e0a69108c8e5c8d190569f6ad565c439807a6b72

`thrift_library` is a macro where you give it the .thrift sources and a list of the languages to generate bindings for, and the macro expands to multiple genrules to run the thrift compiler along with appropriate language-specific targets that refer to the generated code.

A macro is "just" a Starlark function defined in a bzl file. You can call it from BUCK and it can call rules.

So in your case the user might write:

```starlark
load("@...//defs.bzl", "proto_library")

proto_library(
    name = "googlesearch",
    src = ["googlesearch.proto"],
    languages = ["rust", "cxx"],
)
```

and `proto_library` is something resembling:

```starlark
def proto_library(name, src, languages, deps = [], visibility = []):
    for lang in languages:
        if lang == "rust":
            ...
        elif lang == "cxx":
            cxx_proto_library(
                name = "{}-cxx".format(name),
                src = src,
                deps = deps,
                visibility = visibility,
            )
        else:
            fail(...)

def cxx_proto_library(name, src, deps = [], visibility = []):
    protoc_generate_sources(  # your own rule as you've already done, or just the builtin genrule
        name = "{}-generated".format(name),
        ...
        visibility = [],
    )

    cxx_library(
        name = name,
        srcs = [
            ":{}-generated[pb.cc]".format(name),
        ],
        headers = [
            ":{}-generated[pb.h]".format(name),
        ],
        visibility = visibility,
    )
```

Macros run during "parsing" which is when buck runs a BUCK file to figure out what targets are in it. So for example a `fail` located in a macro would get hit during `buck2 targets :`.

Rules run during "analysis" which is when buck wires up a graph with all the dependency relationships between all the targets. For example a `fail` located in a rule would _not_ get hit during `buck2 targets :`, but _would_ get hit during `buck2 cquery :`.

Then obviously there's "building" where your actual actions get run; a `compile_error!` inside Rust code would get hit during `buck2 build` but not `targets` / `cquery`.

That's my mental model anyway. These phases are not as distinct as they sound but it's helpful to understand macros vs rules.

**Comment by benbrittain** [comment]
```
        srcs = [
            ":{}-generated[pb.cc]".format(name),
        ],
```

This was working great until I tried to return multiple `.pb.cc` files in a DefaultOutput.

where upon I got a

```
Expected a single artifact from root//gen_proto:protos_cxx-generated[pb.cc] 
(prelude//platforms:default#524f8da68ea2a374), but it returned 2 artifacts
```

Is there a way to return more than one generated source file/artifact using this pattern?

**Comment by dtolnay** [comment]
Yes, they either need to be `sub_targets`:

```starlark
return [DefaultInfo(
    default_output = None,
    sub_targets = {
        k: [DefaultInfo(default_output = v)]
        for k, v in what_i_built.items()
    },
)]
```

or you return a directory:

```starlark
output_dir ctx.actions.declare_output("...", dir = True)
...
return [DefaultInfo(default_output = output_dir)]
```

**Comment by cjhopman** [comment]
If you need to consume it in something like `srcs`, you'll have the challenge that either you need to explicitly list out each of the subtargets or that srcs will somehow need to be directory-aware (i.e. it would need to know what to do if one of its srcs is a directory). Most things won't be directory-aware in that way and you'd probably want the subtargets and then need to list out each one at the consumer.

If writing your own rules, an alternative to making srcs directory-aware would be to make them support a particular provider that you produce with a list of srcs instead of using the default attr.src() conversion.

**Comment by dtolnay** [comment]
`rust_library`'s `srcs` is directory aware.

https://github.com/facebook/buck2/commit/ae46e4c3745783727dff40364eadc4e0c9425726

```starlark
rust_library(
    ...
    mapped_srcs: {
        ":this_produces_directory": "src",
        ":this_produces_subtargets[foo.rs]": "src/foo.rs",
    },
)
```

**Comment by cjhopman** [comment]
Yeah, I expect that's correct that if you want to produce an unknown number of output srcs you'd need to write a new rule or change the prelude to support it.

Internally for protobuf, we derive the list of expected outputs based on the list of inputs srcs so that we can construct a cxx_library that explicitly lists out those srcs. So the (very basic and probably with some details wrong) idea is:

```
def codegen_library(name, srcs):
   cpp_outputs = [s + ".cpp" for s in srcs]
   h_outputs = [s + ".hpp" for s in srcs]
   gen_name = name + "-gen"
   genrule(
     name = gen_name,
     outs = {p: p for p in cpp_outputs + h_outputs},
     cmd = <...>
   )
   cpp_library(
     name = name + "-cpp",
     headers = ["{}[{}]".format(gen_name, h) for h in h_outputs],
     srcs = ["{}[{}]".format(gen_name, c) for c in cpp_outputs],
   )
```

**Comment by benbrittain** [comment]
Unfortunately, the above quickly didn't meet our needs. While it works great for small isolated protobuf files, once you need to rely on protobufs that use `import` statements from other locations, the genrule approach quickly doesn't scale.

**Comment by cjhopman** [comment]
I think you should be able to extend it. To start (i'll propose something better beloow): internally for protobuf we have something like that but with one added piece to support deps: we add a merge_tree genrule (with `name=name + "-protoc-includes"` or so) that for each protobuf library will produce a directory of the merge of its own srcs and the "-protoc-includes" dirs of each of its deps, and then pass that dir to protoc with `-I`. So each protobuf_library ends up constructing a dir with all of its transitive deps of proto files.

Now, we need to do it that awkward way because it needs to work with buck1. I think you could do it much better if you would use a single new rule for invoking protoc and passing along that includes information via providers rather than having to pass it all along by like generated genrule names and the filesystem. So, you'd have a rule for invoking protoc, but you'd still have the cxx_library like you do with the genrule and your protoc would have all of the .c/.h accessible via subtargets like you did with the genrule.

**Comment by benbrittain** [comment]
I think I can put together something similar to what you all do with the `protoc-includes`. :+1: 

>  think you could do it much better if you would use a single new rule for invoking protoc and passing along that includes information via providers 

This was what I tried at first, but we didn't seem to have a way of taking that collected information and feeding it into the cxx_library rule. That's what the above discussion on variable inputs to cxx_library was talking about. That's when we tried using `anon_target` but ran into #215.

**Comment by cjhopman** [comment]
> This was what I tried at first, but we didn't seem to have a way of taking that collected information and feeding it into the cxx_library rule.

I think you misunderstood just a bit. You would continue to use the macro to expand this into two targets, a protoc_library (or whatever name) and a cxx_library (just like you latest example produces a genrule and a cxx_library target). The protoc_library would then return in its providers something like:

```
  return [DefaultInfo(subtargets = {
     "something.h": DefaultInfo(default_outputs = [something_h_artifact]),
     "something.c": DefaultInfo(default_outputs = [something_c_artifact]),
      ...
   }
 ```
 
 You would still need to be able to compute all those subtarget names statically so that they could be specified as srcs/headers to the cxx_library (like you do with the genrule example).

**Comment by benbrittain** [comment]
Thanks for clarifying! That made total sense and was really quick to implement. I appreciate the help

**Comment by alexlian** [comment]
Looking like this is resolved per comments, so closing.
